aten/src/ATen/templates/RegisterDispatchKey.cpp

```diff
@@ -50,6 +50,15 @@
 $dispatch_headers
 $ops_headers
 
+#ifndef EAGER_REGISTRATION
+#define EAGER_REGISTRATION 1
```

Comment:
I'm a little confused by this split between `DispatchKeyNativeFunctions.h` and `RegisterDispatchKey.cpp`.

Don't we always end up `#include`ing `DispatchKeyNativeFunctions.h` in this file, so the `Register${BackendName}${DispatchKey}Modules` function will already be declared? (and the macro above will already be defined).

Reply:
Maybe in practice this might be true. But its not explicitly clear just from the template definition that this is the case. Which is why I added the `#ifndef` clause. If it happens that there is a case (either now or in the future) where the `DispatchKeyNativeFunctions.h` is not included, it would avoid a compile error here and revert back to the default behaviour.

Reply:
I think the "delay registration using a lambda" feels like a niche enough use case to me that we should try to implement it with less clutter to this file. Instead of using the macros, how do you feel about unconditionally generating the `Register{...}Modules` lambda? And still using `${eager_registration}` to decide in the codegen whether or not we call it directly in the file or not (to run the registrations at load time).

The `initTSBackend()` function wouldn't need to check a macro - it would just always call the lambda to run the registrations. You might run into problems if we got the codegen wrong and accidentally registered the TS backend on startup, but you could also just add a unit test for this (assert in one of the C++ tests that there aren't any kernels registered to the Lazy key on startup).

Reply:
yep, that sounds reasonable.

I kind of don't want to clutter up this file with the extra macros, mostly because this is a pretty core template used in a bunch of different contexts (used for most of our eager mode dispatcher registrations too) and so anyone looking at this file for non-LTC purposes would now have to reason about the macros.

@wconstab would you rather make changes to this PR directly, or address feedback later?

Reply:
I think we can make the changes here.  I already landed this as-is on the staging branch, but we are almost at the point of abandoning the staging branch so I think we can start to do things differently here instead of striving to stay in sync there.

It also depends how we make the changes; if you have some clear ideas of what i should change I can just make the changes.  If you actually want to try to modify it yourself, I could either land it first and you change it, or maybe you could commandeer this diff.

Reply:
I think what @antoniojkim put sounds reasonable:

(1) kill the macro and unconditionally generate the `Register${BackendName}${DispatchKey}Modules();`

(2) In the codegen, if the `eager_registration` argument is true, then emit the code to run the function immediately (which this PR already does), and then also set it to re-assign it to a lambda that no-ops (that way you won't accidentally call it and register things multiple times - although I think the risk of this being a problem in practice is pretty small, since nobody should ever call this function except for the torchscript one. So if this is hard to do then I think it's ok if you don't add it).